raiden/raiden_event_handler.py

```diff
@@ -295,13 +304,76 @@ def handle_contract_send_channelunlock(
             raiden: RaidenService,
             channel_unlock_event: ContractSendChannelBatchUnlock,
     ):
-        channel = raiden.chain.payment_channel(
+        payment_channel: PaymentChannel = raiden.chain.payment_channel(
             channel_unlock_event.token_network_identifier,
             channel_unlock_event.channel_identifier,
         )
+        token_network: TokenNetwork = payment_channel.token_network
+
+        # Fetch on-chain balance hashes for both participants
+        participants_details = token_network.detail_participants(
+            raiden.address,
+            channel_unlock_event.participant,
+            channel_unlock_event.channel_identifier,
+        )
+
+        our_details = participants_details.our_details
+        our_locksroot = our_details.locksroot
+
+        partner_details = participants_details.partner_details
+        partner_locksroot = partner_details.locksroot
+        record = None
+
+        if (partner_details.address == channel_unlock_event.participant and
+                partner_locksroot != EMPTY_HASH):
+            # Partner account
+            record = raiden.wal.storage.get_latest_state_change_by_data_field({
+                'balance_proof.chain_id': raiden.chain.network_id,
+                'balance_proof.token_network_identifier': to_checksum_address(
+                    channel_unlock_event.token_network_identifier,
+                ),
+                'balance_proof.channel_identifier': channel_unlock_event.channel_identifier,
+                'balance_proof.sender': to_checksum_address(
+                    participants_details.partner_details.address,
+                ),
+                'balance_proof.locksroot': serialize_bytes(partner_locksroot),
+            })
+        elif (our_details.address == channel_unlock_event.participant and
+                our_locksroot != EMPTY_HASH):
+            # Our account
+            record = raiden.wal.storage.get_latest_event_by_data_field({
+                'balance_proof.chain_id': raiden.chain.network_id,
+                'balance_proof.token_network_identifier': to_checksum_address(
+                    channel_unlock_event.token_network_identifier,
+                ),
+                'balance_proof.locksroot': serialize_bytes(our_locksroot),
+                'channel_identifier': channel_unlock_event.channel_identifier,
+            })
+        else:
+            raise RaidenUnrecoverableError(
+                "Failed to find state/event that match current channel locksroots",
+            )
+
+        # Replay state changes until a channel state is reached where
+        # this channel state has the participants balance hash.
+        restored_channel_state = channel_state_until_state_change(
```

Comment:
Note: This may raise an unrecoverable error if the channel doesn't exist. Which is fine because:

- either something is wrong in the query, therefor is a bug
- we called this handler twice, which is a bug